From ca1f235f6e351187a379e9c46a3e42889c27148f Mon Sep 17 00:00:00 2001
From: Karim Pinter <karim.pinter@qt.io>
Date: Tue, 8 Jul 2025 14:40:48 +0300
Subject: [PATCH 185/239] Set default stack size safety margin for VxWorks

The default safety margin for VxWorks is 1/8 of the stack size, which is read
from the OS. Also adds environment variable QV4_STACK_SOFT_LIMIT, to
overwrite V4 stack safety margin.
These can be used to finetune the application with limited resources, also on
other OSes.

Task-number: QTBUG-115777
Change-Id: I82b5b02e75fae3f5d1971504d16a3dc0f1d5f3fd
Pick-to: 6.9 6.8
Reviewed-by: Ulf Hermann <ulf.hermann@qt.io>
(cherry picked from commit 3192915698fa5f383f37055df854a39161852506)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/qml/doc/src/javascript/finetuning.qdoc |  7 +++++
 src/qml/jsruntime/qv4engine.cpp            | 16 ++++++-----
 src/qml/jsruntime/qv4engine_p.h            | 16 ++++++++---
 src/qml/memory/qv4stacklimits.cpp          | 31 +++-------------------
 src/qml/memory/qv4stacklimits_p.h          | 24 +++++++++++++++++
 5 files changed, 58 insertions(+), 36 deletions(-)

diff --git a/src/qml/doc/src/javascript/finetuning.qdoc b/src/qml/doc/src/javascript/finetuning.qdoc
index 2f0fdac4ca..cc04c5f11c 100644
--- a/src/qml/doc/src/javascript/finetuning.qdoc
+++ b/src/qml/doc/src/javascript/finetuning.qdoc
@@ -38,6 +38,13 @@ Running JavaScript code can be influenced by a few environment variables, partic
             handle an excessive number of objects at the same time, this stack might overrun.
             If it contains a number, this environment variable is interpreted as the size in bytes
             of the memory area that will be allocated as the stack for the garbage collector.
+    \row
+        \li \c{QV4_STACK_SOFT_LIMIT}
+        \li When this environment variable is set, the JavaScript engine will throw a "RangeError:
+            Maximum call stack size exceeded" exception once the call stack usage reaches the
+            defined soft limit. If the variable is not set, the engine defaults to runtime detected
+            or Qt predefined limits depending on the OS.
+
     \row
         \li \c{QV4_CRASH_ON_STACKOVERFLOW}
         \li Usually the JavaScript engine tries to catch C++ stack overflows caused by
diff --git a/src/qml/jsruntime/qv4engine.cpp b/src/qml/jsruntime/qv4engine.cpp
index d7216c11d7..2a2901e8a9 100644
--- a/src/qml/jsruntime/qv4engine.cpp
+++ b/src/qml/jsruntime/qv4engine.cpp
@@ -118,6 +118,7 @@ int ExecutionEngine::s_maxCallDepth = -1;
 int ExecutionEngine::s_jitCallCountThreshold = 3;
 int ExecutionEngine::s_maxJSStackSize = 4 * 1024 * 1024;
 int ExecutionEngine::s_maxGCStackSize = 2 * 1024 * 1024;
+int ExecutionEngine::s_stackSizeSoftLimit = -1;
 
 ReturnedValue throwTypeError(const FunctionObject *b, const QV4::Value *, const QV4::Value *, int)
 {
@@ -308,6 +309,12 @@ void ExecutionEngine::initializeStaticMembers()
     if (ok && envMaxGCStackSize > 0)
         s_maxGCStackSize = envMaxGCStackSize;
 
+    const int envStackSizeSoftLimit = qEnvironmentVariableIntValue("QV4_STACK_SOFT_LIMIT", &ok);
+    if (ok && envStackSizeSoftLimit > -1)
+        s_stackSizeSoftLimit = envStackSizeSoftLimit;
+    else
+        s_stackSizeSoftLimit = -1;
+
     if (qEnvironmentVariableIsSet("QV4_CRASH_ON_STACKOVERFLOW")) {
         s_maxCallDepth = std::numeric_limits<qint32>::max();
     } else {
@@ -361,13 +368,10 @@ ExecutionEngine::ExecutionEngine(QJSEngine *jsEngine)
         }
     }
 
-    if (s_maxCallDepth < 0) {
-        const StackProperties stack = stackProperties();
-        cppStackBase = stack.base;
-        cppStackLimit = stack.softLimit;
-    } else {
+    if (s_maxCallDepth < 0)
+        setCppStackProperties();
+    else
         callDepth = 0;
-    }
 
     // We allocate guard pages around our stacks.
     const size_t guardPages = 2 * WTF::pageSize();
diff --git a/src/qml/jsruntime/qv4engine_p.h b/src/qml/jsruntime/qv4engine_p.h
index ab5de81a9a..a15c39e231 100644
--- a/src/qml/jsruntime/qv4engine_p.h
+++ b/src/qml/jsruntime/qv4engine_p.h
@@ -826,6 +826,16 @@ private:
 #endif
     }
 
+    void setCppStackProperties()
+    {
+        const StackProperties stack = stackProperties();
+        cppStackBase = stack.base;
+        if (s_stackSizeSoftLimit == -1)
+            cppStackLimit = stack.softLimit;
+        else
+            cppStackLimit = incrementStackPointer(stack.base, s_stackSizeSoftLimit);
+    }
+
     bool hasCppStackOverflow()
     {
         if (s_maxCallDepth >= 0)
@@ -836,9 +846,8 @@ private:
 
         // Double check the stack limits on failure.
         // We may have moved to a different thread.
-        const StackProperties stack = stackProperties();
-        cppStackBase = stack.base;
-        cppStackLimit = stack.softLimit;
+        setCppStackProperties();
+
         return !inStack(currentStackPointer());
     }
 
@@ -856,6 +865,7 @@ private:
     static int s_jitCallCountThreshold;
     static int s_maxJSStackSize;
     static int s_maxGCStackSize;
+    static int s_stackSizeSoftLimit;
 
 #if QT_CONFIG(qml_debug)
     QScopedPointer<QV4::Debugging::Debugger> m_debugger;
diff --git a/src/qml/memory/qv4stacklimits.cpp b/src/qml/memory/qv4stacklimits.cpp
index 4c2cfbea74..5e43900069 100644
--- a/src/qml/memory/qv4stacklimits.cpp
+++ b/src/qml/memory/qv4stacklimits.cpp
@@ -71,35 +71,11 @@ enum StackDefaults : qsizetype {
 #endif
 };
 
-// We may not be able to take the negative of the type
-// used to represent stack size, but we can always add
-// or subtract it to/from a quint8 pointer.
-
-template<typename Size>
-static void *incrementStackPointer(void *base, Size amount)
-{
-#if Q_STACK_GROWTH_DIRECTION > 0
-    return static_cast<quint8 *>(base) + amount;
-#else
-    return static_cast<quint8 *>(base) - amount;
-#endif
-}
-
-template<typename Size>
-static void *decrementStackPointer(void *base, Size amount)
-{
-#if Q_STACK_GROWTH_DIRECTION > 0
-    return static_cast<quint8 *>(base) - amount;
-#else
-    return static_cast<quint8 *>(base) + amount;
-#endif
-}
-
-static StackProperties createStackProperties(void *base, qsizetype size = PlatformStackSize)
+static StackProperties createStackProperties(void *base, qsizetype size = PlatformStackSize, qsizetype margin = PlatformSafetyMargin)
 {
     return StackProperties {
         base,
-        incrementStackPointer(base, size - PlatformSafetyMargin),
+        incrementStackPointer(base, size - margin),
         incrementStackPointer(base, size),
     };
 }
@@ -242,7 +218,8 @@ StackProperties stackProperties()
 {
     TASK_DESC taskDescription;
     taskInfoGet(taskIdSelf(), &taskDescription);
-    return createStackProperties(taskDescription.td_pStackBase, taskDescription.td_stackSize);
+    return createStackProperties(taskDescription.td_pStackBase, taskDescription.td_stackSize,
+                                 taskDescription.td_stackSize / 8);
 }
 
 #else
diff --git a/src/qml/memory/qv4stacklimits_p.h b/src/qml/memory/qv4stacklimits_p.h
index 9e7fdb279b..a87b541ed9 100644
--- a/src/qml/memory/qv4stacklimits_p.h
+++ b/src/qml/memory/qv4stacklimits_p.h
@@ -29,6 +29,30 @@ QT_BEGIN_NAMESPACE
 
 namespace QV4 {
 
+// We may not be able to take the negative of the type
+// used to represent stack size, but we can always add
+// or subtract it to/from a quint8 pointer.
+
+template<typename Size>
+static const void *incrementStackPointer(const void *base, Size amount)
+{
+#if Q_STACK_GROWTH_DIRECTION > 0
+    return static_cast<const quint8 *>(base) + amount;
+#else
+    return static_cast<const quint8 *>(base) - amount;
+#endif
+}
+
+template<typename Size>
+static void *decrementStackPointer(void *base, Size amount)
+{
+#if Q_STACK_GROWTH_DIRECTION > 0
+    return static_cast<quint8 *>(base) - amount;
+#else
+    return static_cast<quint8 *>(base) + amount;
+#endif
+}
+
 // Note: This does not return a completely accurate stack pointer.
 //       Depending on whether this function is inlined or not, we may get the address of
 //       this function's stack frame or the caller's stack frame.
-- 
2.51.2

